runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java

```diff
@@ -370,7 +373,7 @@ boolean bloomFilterMightContain(RandomAccessData keyBytes) {
     position(rawChannel, footer.getBloomFilterPosition());
     bloomFilter = ScalableBloomFilterCoder.of().decode(Channels.newInputStream(rawChannel));
 
-    indexPerShard = new HashMap<>();
+    indexPerShard = new ConcurrentHashMap<>();
```

Comment:
Note that initializeBloomFilterAndIndexPerShard is synchronized which is the only method that mutates the indexPerShard variable. Since initializeBloomFilterAndIndexPerShard is invoked within overKeyComponents that means there is a memory barrier between the thread that created the indexPerShard and other threads ensuring that the object is safely published across threads.

Also, the [Javadoc for HashMap](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html) says that concurrent reads are ok, its concurrent modifications that require synchronization:
"If multiple threads access a hash map concurrently, and at least one of the threads modifies the map structurally, it must be synchronized externally."

Reply:
`indexPerShard` is updated at [initializeForKeyedRead](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L562), which is not synchronized. 

#### IllegalStateException
Let's assume that thread T1 and T2 calls `overKeyComponents` and enter `initializeForKeyedRead` for shardId `K` at the same time. And `indexPerShard` doesn't have `K`:

1. T1 checks [`indexPerShard.containsKey(K)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L495) and it returns `false`. 
2. T2 checks [`indexPerShard.containsKey(K)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L495) and it returns `fase`.
3. T1 advances quickly and invokes [`indexPerShard.put(K, ...)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L562). Now, `indexPerShard` has `K`.
2. T2 checks [`checkState(indexPerShard.get(K) == null)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L500-L503). It throws `IllegalStateException`.

#### NullPointerException
Let's assume that thread T1 and T2 calls `overKeyComponents`. T1 is for shardId `K1` while T2 is shardId `K2`, and `indexPerShard` has K1:

1. T1 returns from `intializeForKeyedRead` as [`indexPerShard.containsKey(K1)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L495) is `true`.
2. T1 stops at [`indexPerShard.get(K1)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L210)  at `overKeyComponents`.
3. T2 advances and invokes [`indexPerShard.put(K2, ...)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L562) at `intializeForKeyedRead`.
4. T1 can get `null` from `indexPerShard.get(K1)`, and will throw `NullPointerException` as it tries to invoke [`floorKey`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L210).

My theory on this is that 
* T2 is in [`HashMap.rehash()`](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Hashtable.java#l389) where a `newMap` is created and then entries in `oldMap` are copied to `newMap`.
* T1 accesses `newMap` before copy finishes. So, it sees `null` for `K1` which exists in `oldMap` but not in `newMap`.

Reply:
Actually, any Map supporting thread-safe put/get is ok for `indexPerShard` (e.g. `Hashtable`).

Reply:
After writing `IllegalStateException` scenario above, I realized that changing the type of `indexPerShard` to a thread-safe Map cannot prevent `IllegalStateException` at [`checkState(indexPerShard.get(K) == null)`](https://github.com/apache/beam/blob/v2.38.0-RC1/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/IsmReaderImpl.java#L500-L503).`.

I added additional [commit](https://github.com/apache/beam/pull/17201/commits/b74fc99bee7e1aa11b38495d53ab0b3147c17b34) to remove the check. It's not necessary as the condition is checked at if-clause just before it.

In addition, as `initializeForKeyedRead` allows concurrent access, it has a benefit to update `indexPerShard` for multiple shardIds, but it can update a key multiple times if multiple threads try to update the key at the same time. We can fine-control the synchronization, but it may make the code a little bit more complex (e.g. using `computeIfAbsent` with `indexPerShard`). Do you think we should do more optimization?

Reply:
Thanks for the thorough explanation. Using computeIfAbsent is limited in concurrency since it can only update keys that are in different hash buckets but I/O is so expensive so it should be worthwhile to do that swap.